### PR TITLE
issue: 1228170 Fix enable outgoing RST

### PR DESCRIPTION
**Pull request by serglelv**
Fix incoming packet that has the ACK bit set but not the SYN bit set

### CODE REVIEW

**Comment by serglelv** [comment]
@liranoz12 Hi Liran, please review. 
Please note TCP_EVENT_CLONE_PCB is used due to unexpected tcp_close events and race conditions with PD testing.

File: src/vma/lwip/tcp_in.c [hunk 1]
```diff
@@ -341,19 +341,36 @@ tcp_listen_input(struct tcp_pcb_listen *pcb, tcp_in_data* in_data)
...
+      ptcp_hdr->seqno = htonl(in_data->ackno);
+      ptcp_hdr->ackno = 0;
+      TCPH_HDRLEN_FLAGS_SET(ptcp_hdr, TCP_HLEN/4, TCP_RST);
+      ptcp_hdr->wnd = 0;
+      ptcp_hdr->chksum = 0;
+      ptcp_hdr->urgp = 0;
+      npcb->callback_arg = (void*)ptcp_hdr;
+      TCP_DST_NC_SEND(pcb, npcb, ERR_OK, rc);
+    }
   } else if (in_data->flags & TCP_SYN) {
```
Comment (liranoz12) [inline src/vma/lwip/tcp_in.c]: Most of the logic is duplicated from tcp_rst.
why can't we create a temporary pcb and call tcp_rst with it to send the data ?

File: src/vma/lwip/tcp_out.c [hunk 2]
```diff
@@ -1354,6 +1354,8 @@ tcp_rst(u32_t seqno, u32_t ackno, u16_t local_port, u16_t remote_port, struct tc
   /* external_ip_output(p, NULL, local_ip, remote_ip, TCP_TTL, 0, IP_PROTO_TCP) */;
   tcp_tx_pbuf_free(pcb, p);
   LWIP_DEBUGF(TCP_RST_DEBUG, ("tcp_rst: seqno %"U32_F" ackno %"U32_F".\n", seqno, ackno));
+
+  return ERR_OK;
 }
```
Comment (liranoz12) [inline src/vma/lwip/tcp_out.c]: We can remove all tcp_rst changes if we create the temporary pcb object before

File: src/vma/lwip/tcp_in.c [hunk 3]
```diff
@@ -341,19 +341,36 @@ tcp_listen_input(struct tcp_pcb_listen *pcb, tcp_in_data* in_data)
     return ERR_OK;
   }
 
-  if (in_data->flags & TCP_FIN) {
-    /* An incoming FIN should be ignored. Return. */
-    return ERR_OK;
-  }
-
```
[the remark this answers is not in the document]
Comment (serglelv) [reply src/vma/lwip/tcp_in.c]: It provides that a TCP segments with FIN-ACK in the LISTEN state does  trigger the sending of RST.  The same behaviour was observed for Linux as well.

File: src/vma/lwip/tcp_in.c (hunk 1 above)
Comment (serglelv) [reply src/vma/lwip/tcp_in.c]: I agree, it can be optimized. Thanks.

**Comment by serglelv** [comment]
@liranoz12 Hi Liran, please review.

File: src/vma/lwip/tcp_in.c [hunk 4]
```diff
@@ -341,19 +341,24 @@ tcp_listen_input(struct tcp_pcb_listen *pcb, tcp_in_data* in_data)
...
+    if (npcb == NULL) {
+      LWIP_DEBUGF(TCP_DEBUG, ("tcp_listen_input: could not allocate PCB\n"));
+      TCP_STATS_INC(tcp.memerr);
+      return ERR_MEM;
+    }
+    ip_addr_copy(npcb->local_ip, in_data->iphdr->dest);
+    ip_addr_copy(npcb->remote_ip, in_data->iphdr->src);
+    npcb->remote_port = in_data->tcphdr->src;
+    TCP_DST_NC_SEND(pcb, npcb, ERR_OK, rc);
+    tcp_rst(in_data->ackno, 0, in_data->tcphdr->dest, in_data->tcphdr->src, npcb, &flags);
```
Comment (liranoz12) [inline src/vma/lwip/tcp_in.c]: I dont see here any free/cleanup for the newpcb.
Maybe the valgrind is realted to this...

File: src/vma/lwip/tcp_out.c [hunk 5]
```diff
@@ -1343,8 +1343,13 @@ tcp_rst(u32_t seqno, u32_t ackno, u16_t local_port, u16_t remote_port, struct tc
   tcphdr->dest = htons(remote_port);
   tcphdr->seqno = htonl(seqno);
   tcphdr->ackno = htonl(ackno);
-  TCPH_HDRLEN_FLAGS_SET(tcphdr, TCP_HLEN/4, TCP_RST | TCP_ACK);
-  tcphdr->wnd = PP_HTONS(( TCP_WND  & 0xFFFF ));
+  if (!flags) {
```
Comment (liranoz12) [inline src/vma/lwip/tcp_out.c]: it is better to pass the value of the flags

File: src/vma/lwip/tcp_out.c [hunk 6]
```diff
@@ -1343,8 +1343,13 @@ tcp_rst(u32_t seqno, u32_t ackno, u16_t local_port, u16_t remote_port, struct tc
...
+      tcphdr->wnd = PP_HTONS(( TCP_WND  & 0xFFFF ));
+      TCPH_HDRLEN_FLAGS_SET(tcphdr, TCP_HLEN/4, TCP_RST | TCP_ACK);
+  } else {
+      tcphdr->wnd = 0;
+      TCPH_HDRLEN_FLAGS_SET(tcphdr, TCP_HLEN/4, *flags);
+  }
```
Comment (liranoz12) [inline src/vma/lwip/tcp_out.c]: tcphdr->wnd = 0; is this unique behavior for this case ? or all reset should be sent we wnd = 0

File: src/vma/sock/sockinfo_tcp.cpp [hunk 7]
```diff
@@ -2982,6 +2983,34 @@ err_t sockinfo_tcp::syn_received_drop_lwip_cb(void *arg, struct tcp_pcb *newpcb,
...
+
+	ASSERT_LOCKED(listen_sock->m_tcp_con_lock);
+	listen_sock->m_tcp_con_lock.unlock();
+
+	new_sock->set_conn_properties_from_pcb();
+	new_sock->create_dst_entry();
+
+	if (new_sock->m_p_connected_dst_entry) {
+		new_sock->prepare_dst_to_send(false);
+		new_sock->abort_connection();
```
Comment (liranoz12) [inline src/vma/sock/sockinfo_tcp.cpp]: abort_connection can send reset if the socket state != closed.
can we use it ?

File: src/vma/sock/sockinfo_tcp.cpp [hunk 8]
```diff
@@ -2982,6 +2983,34 @@ err_t sockinfo_tcp::syn_received_drop_lwip_cb(void *arg, struct tcp_pcb *newpcb,
...
+	}
+	close(new_sock->get_fd());
+
+	listen_sock->m_tcp_con_lock.lock();
+
+	return ERR_OK;
+}
```
Comment (liranoz12) [inline src/vma/sock/sockinfo_tcp.cpp]: The logic is very similar to : syn_received_lwip_cb / syn_received_drop_lwip_cb
Can we avoid this code duplication ?

File: src/vma/lwip/tcp_in.c (hunk 4 above)
Comment (serglelv) [reply src/vma/lwip/tcp_in.c]: Implicit deallocation is used. This code cannot be "traced" with Valgrind using sockperf.

File: src/vma/lwip/tcp_out.c (hunk 6 above)
Comment (serglelv) [reply src/vma/lwip/tcp_out.c]: The behavior was aligned with Linux (i.e. wnd=0) in LISTEN state < S. (SYN-ACK)

File: src/vma/lwip/tcp_out.c (hunk 5 above)
Comment (serglelv) [reply src/vma/lwip/tcp_out.c]: In this case it's probably better than to pass the value.

**Comment by liranoz12** [comment]
bot:retest

**Pull request closed without merging**